pkg/controller/master/addon/common.go

```diff
@@ -0,0 +1,218 @@
+			Repo:          aObj.Spec.Repo,
+			ValuesContent: vals,
+			Version:       aObj.Spec.Version,
+		},
+	}
+	_, err = h.helm.Create(hc)
+	if err != nil {
+		return fmt.Errorf("error creating helmchart object %v", err)
+	}
+
+	return nil
+}
+
+func defaultValues(a *harvesterv1.Addon) (string, error) {
+	if a.Spec.ValuesContent != "" {
+		return a.Spec.ValuesContent, nil
+	}
+
+	valsEncoded, ok := a.Annotations[util.AddonValuesAnnotation]
+	if ok {
+		valByte, err := base64.StdEncoding.DecodeString(valsEncoded)
+		if err != nil {
+			return "", fmt.Errorf("error decoding addon defaults: %v", err)
+		}
+
+		return string(valByte), nil
+	}
+	// no overrides. Use packaged chart defaults
+	return "", nil
+}
+
+func isJobComplete(j *batchv1.Job) bool {
+	if j.Status.CompletionTime != nil {
+		return true
+	}
+
+	for _, v := range j.Status.Conditions {
+		if v.Type == batchv1.JobFailed && v.Reason == "BackoffLimitExceeded" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (h *Handler) isAddonUpdateNeeded(aObj *harvesterv1.Addon) (bool, error) {
+	hc, owned, err := h.getAddonHelmChart(aObj)
+	if err != nil {
+		return false, err
+	}
+
+	if hc == nil || !owned {
+		return true, nil
+	}
+
+	// e.g. downstream helmchart was manually changed, addon will re-sync
+	need, _, err := h.isHelmchartUpdateNeeded(aObj, hc)
+
+	return need, err
+}
+
+func (h *Handler) enqueueAfter(aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	h.addon.EnqueueAfter(aObj.Namespace, aObj.Name, enqueueInterval*time.Second)
+	return aObj, nil
+}
+
+func (h *Handler) getJob(hc *helmv1.HelmChart) (*batchv1.Job, error) {
+	return h.job.Cache().Get(hc.Namespace, hc.Status.JobName)
+}
+
+func (h *Handler) currentDeletionJob(hc *helmv1.HelmChart) (*batchv1.Job, bool, error) {
+	j, err := h.getJob(hc)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// job creation timestamp should be after deletion timestamp of
+	// helm chart to ensure that we are checking the correct job
+	if j.CreationTimestamp.After(hc.DeletionTimestamp.Time) || j.CreationTimestamp.Equal(hc.DeletionTimestamp) {
+		return j, true, nil
+	}
+
+	return j, false, nil
+}
+
+func (h *Handler) currentInstallationJob(hc *helmv1.HelmChart, a *harvesterv1.Addon) (*batchv1.Job, bool, error) {
+	logrus.Debugf("querying current installation job for addon %s", a.Name)
+
+	j, err := h.getJob(hc)
+	if err != nil {
+		return nil, false, err
+	}
+
+	lastUpdatedTime, err := time.Parse(time.RFC3339, harvesterv1.AddonOperationInProgress.GetLastUpdated(a))
+	if err != nil {
+		return nil, false, fmt.Errorf("error parsing last updated time for AddonOperationInProgress: %v", err)
+	}
+
+	metav1LastUpdatedTime := metav1.NewTime(lastUpdatedTime)
+	logrus.Debugf("last updated time on the addon: %s", lastUpdatedTime)
+	// job creation timestamp should be after the last updated time stamp
+	// on the inprogress condition
+	if j.CreationTimestamp.After(lastUpdatedTime) || j.CreationTimestamp.Equal(&metav1LastUpdatedTime) {
```

Comment:
when addon is upgraded.. it will be moved from `Deployed` to `Updating` status.. and `markInProgress` will be set again. This will update the timestamps on conditions.

When HelmChart is updated with new values another helm install job is created with a newer timestamp which will be equal to or after the addon update is triggered.